src/xutils.cpp

```diff
@@ -0,0 +1,89 @@
+/***************************************************************************
+* Copyright (c) 2018, Martin Renou, Johan Mabille, Sylvain Corlay and      *
+* Wolf Vollprecht                                                          *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#include <string>
+#include <vector>
+#include <stdexcept>
+
+#include "xeus/xjson.hpp"
+#include "xeus/xcomm.hpp"
+
+#include "pybind11/pybind11.h"
+
+#include "xutils.hpp"
+
+namespace py = pybind11;
+
+namespace xpyt
+{
+
+    zmq::message_t pybytes_to_zmq_message(py::bytes bytes)
+    {
+        char* buffer;
+        Py_ssize_t length;
+#if PY_MAJOR_VERSION >= 3
+        PyBytes_AsStringAndSize(bytes.ptr(), &buffer, &length);
+#else
+        PyString_AsStringAndSize(bytes.ptr(), &buffer, &length);
+#endif
+        return zmq::message_t(buffer, length);
+    }
+
+    xeus::xjson pyobj_to_xjson(py::object obj)
+    {
+        py::module py_json = py::module::import("json");
+
+        return xeus::xjson::parse(static_cast<std::string>(
+            py::str(py_json.attr("dumps")(obj))
+        ));
+    }
+
+    py::object xjson_to_pyobj(const xeus::xjson& json)
+    {
+        py::module py_json = py::module::import("json");
+
+        return py_json.attr("loads")(json.dump());
+    }
+
+    py::list zmq_buffers_to_pylist(const std::vector<zmq::message_t>& buffers)
+    {
+        py::list bufferlist;
+        for (const zmq::message_t& buffer: buffers)
+        {
+            const char* buf = buffer.data<const char>();
+            bufferlist.attr("append")(py::bytes(buf));
+        }
+        return bufferlist;
+    }
+
+    std::vector<zmq::message_t> pylist_to_zmq_buffers(py::list bufferlist)
+    {
+        std::vector<zmq::message_t> buffers;
+        for (py::handle bytes: bufferlist)
+        {
+            buffers.push_back(pybytes_to_zmq_message(bytes.cast<py::bytes>()));
+        }
+        return buffers;
+    }
+
+    py::object cppmessage_to_pymessage(const xeus::xmessage& msg)
+    {
+        xeus::xjson json_msg;
+        json_msg["header"] = msg.header();
+        json_msg["parent_header"] = msg.parent_header();
+        json_msg["metadata"] = msg.metadata();
+        json_msg["content"] = msg.content();
+
+        py::dict py_msg = xjson_to_pyobj(json_msg);
```

Comment:
Doing

```cpp
py_msg["header"] = xjson_to_pyobj(msg.header());
py_msg["parent_header"] = xjson_to_pyobj(msg.parent_header());
py_msg["metadata"] = xjson_to_pyobj(msg.metadata());
py_msg["content"] = xjson_to_pyobj(msg.content());
```

Could prevent the copy into `json_msg`?

Reply:
Then we can work on a means to streamline nlohmann_json to python json conversion...

Reply:
Actually, (in a different PR), we could override `to_json` and `from_json` for `py::object`. Just like the JSON module of python, we would check if the object have `from_json` and `to_json` overloaded...

The only issue is that we would need to override these in the pybind11 namespace...

Reply:
```c++
py_msg["header"] = xjson_to_pyobj(msg.header());
py_msg["parent_header"] = xjson_to_pyobj(msg.parent_header());
py_msg["metadata"] = xjson_to_pyobj(msg.metadata());
py_msg["content"] = xjson_to_pyobj(msg.content());
```
will only transform one big copy into 4 smaller copies, but at the end of the day it will be the same, am I right?

Reply:
No. This would save a copy.